Source/Libraries/openXDA.APIAuthentication/ImpersonationSecurityProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿//******************************************************************************************************
+//  ImpersonationSecurityProvider.cs - Gbtc
+//
+//  Copyright © 2023, Grid Protection Alliance.  All Rights Reserved.
+//
+//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
+//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
+//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may not use this
+//  file except in compliance with the License. You may obtain a copy of the License at:
+//
+//      http://opensource.org/licenses/MIT
+//
+//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
+//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
+//  License for the specific language governing permissions and limitations.
+//
+//  Code Modification History:
+//  ----------------------------------------------------------------------------------------------------
+//  07/04/2023 - C. Lackner
+//       Generated original version of source code.
+//
+//******************************************************************************************************
+
+using System;
+using GSF.Configuration;
+using GSF.Diagnostics;
+using GSF.Security;
+
+
+namespace openXDA.APIAuthentication
+{
+    /// <summary>
+    /// Middleware to authenticate users of the API via <see cref="APIAccessKey"/>.
+    /// </summary>
+    public partial class APIAuthenticationMiddleware
+    {
+        private class ImpersonationSecurityProvider : AdoSecurityProvider
+        {
+            #region [ Members ]
+            private const string DefaultMessageUserNotDefined = "User \"{0}\" is not defined.";
+            private const string DefaultMessageUserIsDisabled = "User \"{0}\" is disabled.";
+            private const string DefaultMessageUserIsLockedOut = "User \"{0}\" is not locked out.";
+            private const string DefaultMessageUserPasswordExpired = "User \"{0}\" has an expired password or password has not been set.";
+            private const string DefaultMessageUserHasNoRoles = "User \"{0}\" has not been assigned any roles and therefore has no rights. Contact your administrator.";
+            #endregion
+
+            #region [ Constructor ]
+            public ImpersonationSecurityProvider(string username)
+              : base(username, true, false, false)
+            {
+                UserData = new UserData(username);
+                IsUserAuthenticated = Authenticate();
```

Comment:
Convention seems to be that this property should be assigned in the body of the `Authenticate()` method. I think we should do the same thing here, and have the `APIAuthenticationMiddleware` call `Authenticate()` explicitly on the object.